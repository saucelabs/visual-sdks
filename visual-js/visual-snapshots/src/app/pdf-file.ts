export interface LoadedPdfFile {
  readonly path: string;
  readonly pages: number;
  getPage(page: number): Promise<Buffer>;
}
