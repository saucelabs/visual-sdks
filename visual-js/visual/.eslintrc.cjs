module.exports = {
  extends: ['eslint:recommended', 'plugin:@typescript-eslint/recommended'],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: './tsconfig.json',
    tsconfigRootDir: __dirname,
    sourceType: 'module',
  },
  plugins: ['@typescript-eslint'],
  extends: [
    'plugin:@typescript-eslint/recommended',
    'plugin:prettier/recommended',
  ],
  root: true,
  env: {
    node: true,
    jest: true,
  },
  ignorePatterns: ['.eslintrc.js', 'codegen.ts', 'jest.config.js', 'src/graphql/__generated__'],
  rules: {
    '@typescript-eslint/interface-name-prefix': 'off',
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/explicit-module-boundary-types': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    
    // Allow unused vars that start with _
    "@typescript-eslint/no-unused-vars": [
      "warn",
      {
        "argsIgnorePattern": "^_",
        "varsIgnorePattern": "^_",
        "caughtErrorsIgnorePattern": "^_"
      }
    ],
  },
};
