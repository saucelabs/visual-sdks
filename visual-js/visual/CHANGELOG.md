# @saucelabs/visual

## 0.11.0

### Minor Changes

- 0bd1fac: Removed deprecated functionality.

## 0.10.2

### Patch Changes

- e7abe58: expose nativeClipSelector for full page config

## 0.10.1

### Patch Changes

- 7a88f8d: add ignore region selectors for appium native fullpage

## 0.10.0

### Minor Changes

- f797342: Unify result checking behavior into utility
  Add automatic retry mechanism to file uploads for timeouts

## 0.9.0

### Minor Changes

- 8572e1b: scrollElement native fps

## 0.8.3

### Patch Changes

- 5991881: Fill out statuses in sauceVisualResults for cypress
  Bump all dependencies to use new ignore region calculation system

## 0.8.2

### Patch Changes

- ff9cdb2: Rebase Storybook integration onto Playwright

## 0.8.1

### Patch Changes

- 7b1c3cc: Add playwright support

## 0.8.0

### Minor Changes

- 1399b9c: Debugging info when upload is reaching timeout

## 0.7.3

### Patch Changes

- 8a64e27: Update / regen types + update Cypress snapshot upload behavior

## 0.7.2

### Patch Changes

- 3f1e241: Enable merge baselines command

## 0.7.1

### Patch Changes

- d769340: Add command to merge baselines

## 0.7.0

### Minor Changes

- 4bafb87: Change default diffingMethod to BALANCED

## 0.6.6

### Patch Changes

- 58c9211: replace deprecated field in buildStatus query

## 0.6.5

### Patch Changes

- a1f4373: Update types

## 0.6.4

### Patch Changes

- c5c1ee6: Priotize custom id > build id when checking status

## 0.6.3

### Patch Changes

- f15ec66: Type UUID codegen, add return type from sauceVisualCheck

## 0.6.2

### Patch Changes

- 312a7d7: Add WebElement support for GraphQL API, update shared types

## 0.6.1

### Patch Changes

- 926996f: Add Skip support for SAUCE_VISUAL_SKIP

## 0.6.0

### Minor Changes

- c94072c: Remove deprecated API calls webwebdriverSession and webdriverSessionFromArchive

## 0.5.0

### Minor Changes

- 787387a: deprecate some full page options

## 0.4.0

### Minor Changes

- 0384c59: Bump minor after GH migration

### Patch Changes

- ed72709: migrate source to github

## 0.4.0

### Minor Changes

- move to github
