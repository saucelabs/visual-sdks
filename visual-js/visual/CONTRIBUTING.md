
# Sauce Labs Visual client library


## Installation

```sh
yarn install
```

## Regenerate bindings

```sh
yarn copy-schema
yarn gen
yarn build
```

## Test

```sh
yarn test
```
