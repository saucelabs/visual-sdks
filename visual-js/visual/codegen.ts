import type { CodegenConfig } from '@graphql-codegen/cli';

const username = process.env.SAUCE_USERNAME;
const accessKey = process.env.SAUCE_ACCESS_KEY;

const codegenConfig: CodegenConfig = {
  schema: process.env.VISUAL_API_FOR_CODEGEN || `https://${username}:${accessKey}@api.us-west-1.saucelabs.com/v1/visual/graphql`,
  documents: [
    'src/**/*.ts',
    'src/**/*.graphql',
    '!src/**/__generated__/**/*',
  ],
  emitLegacyCommonJSImports: false,
  config: {
    namingConvention: {
      typeNames: 'change-case-all#pascalCase',
      transformUnderscore: true,
    },
    avoidOptionals: {
      field: true,
      inputValue: false,
      object: true,
      defaultValue: true,
    },
    scalars: {
      WebdriverSessionBlob: "string",
      WebdriverElementID: "string",
      Datetime: "Date | string | number",
      UUID: "string",
    }
  },
  generates: {
    './src/graphql/__generated__/': {
      preset: 'client',
      plugins: [],
    },
  },
};

export default codegenConfig;
