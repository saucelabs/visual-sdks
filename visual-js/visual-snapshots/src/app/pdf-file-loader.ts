import { pdf } from "pdf-to-img";
import { LoadedPdfFile } from "./pdf-file.js";

export interface PdfFileLoader {
  loadPdfFile(path: string): Promise<LoadedPdfFile>;
}

export class LibPdfFileLoader implements PdfFileLoader {
  constructor(private readonly _pdf: typeof pdf = pdf) {}

  public async loadPdfFile(pdfFilePath: string): Promise<LoadedPdfFile> {
    const pdfFile = await this._pdf(pdfFilePath, {
      scale: 1,
    });

    return {
      path: pdfFilePath,
      pages: pdfFile.length,
      getPage: (page) => pdfFile.getPage(page),
    };
  }
}
