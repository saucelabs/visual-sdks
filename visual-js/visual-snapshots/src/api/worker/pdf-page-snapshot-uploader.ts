import path from "path";
import { formatString } from "../../utils/format.js";
import { DiffingMethod, VisualApi } from "@saucelabs/visual";
import { PdfFileLoader } from "../../app/pdf-file-loader.js";

export class PdfPageSnapshotUploader {
  constructor(
    private readonly api: VisualApi,
    private readonly pdfFileLoader: PdfFileLoader
  ) {}

  public async uploadPageSnapshot(
    buildId: string,
    pdfFilePath: string,
    pageNumber: number,
    suiteName: string | undefined,
    testNameFormat: string | undefined,
    snapshotNameFormat: string | undefined
  ) {
    const pdfFile = await this.pdfFileLoader.loadPdfFile(pdfFilePath);
    const page = await pdfFile.getPage(pageNumber);

    const filename = path.basename(pdfFile.path);
    const testName = testNameFormat
      ? formatString(testNameFormat, { filename })
      : undefined;

    const snapshotFormat = this.getSnapshotFormat(snapshotNameFormat);
    const snapshotName = formatString(snapshotFormat, {
      filename,
      page: pageNumber,
    });

    await this.uploadImageAndCreateSnapshot(
      filename,
      pageNumber,
      page,
      buildId,
      snapshotName,
      testName,
      suiteName
    );
  }

  private async uploadImageAndCreateSnapshot(
    file: string,
    pageNumber: number,
    snapshot: Buffer,
    buildId: string,
    snapshotName: string,
    testName?: string,
    suiteName?: string
  ) {
    const uploadId = await this.api.uploadSnapshot({
      buildId,
      image: { data: snapshot },
    });

    console.info(
      `[${file}:${pageNumber}] Uploaded image to build ${buildId}: upload id=${uploadId}.`
    );

    await this.api.createSnapshot({
      buildId,
      uploadId,
      name: snapshotName,
      diffingMethod: DiffingMethod.Balanced,
      testName,
      suiteName,
    });

    console.info(
      `[${file}:${pageNumber}] Created a snapshot ${snapshotName} for build ${buildId}.`
    );
  }

  private getSnapshotFormat(format: string | undefined) {
    if (!format) {
      return `page-{page}`;
    }

    // Page number is always required to make the snapshot names unique
    if (!format.includes("{page}")) {
      format = format += "-{page}";
    }

    return format;
  }
}
