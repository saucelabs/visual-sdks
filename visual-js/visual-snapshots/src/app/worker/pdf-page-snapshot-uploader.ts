import {
  buildFileMetadata,
  formatStringWithFileMetadata,
} from "../../utils/format.js";
import { PdfFileLoader } from "../pdf-file-loader.js";
import { VisualSnapshotsApi } from "../../api/visual-snapshots-api.js";
import { logger as defaultLogger } from "../../logger.js";

export class PdfPageSnapshotUploader {
  constructor(
    private readonly visualSnapshotsApi: VisualSnapshotsApi,
    private readonly pdfFileLoader: PdfFileLoader
  ) {}

  public async uploadPageSnapshot(
    buildId: string,
    pdfFilePath: string,
    pageNumber: number,
    suiteNameFormat: string | undefined,
    testNameFormat: string | undefined,
    snapshotNameFormat: string | undefined
  ) {
    const pdfFile = await this.pdfFileLoader.loadPdfFile(pdfFilePath);
    const page = await pdfFile.getPage(pageNumber);

    const pdfFileMetadata = buildFileMetadata(pdfFilePath, pageNumber);
    const testName = testNameFormat
      ? formatStringWithFileMetadata(testNameFormat, pdfFileMetadata)
      : undefined;
    const snapshotName = formatStringWithFileMetadata(
      snapshotNameFormat ? snapshotNameFormat : `page-{page}`,
      pdfFileMetadata
    );
    const suiteName = suiteNameFormat
      ? formatStringWithFileMetadata(suiteNameFormat, pdfFileMetadata)
      : undefined;

    const logger = defaultLogger.child({
      filePath: pdfFilePath,
      pageNumber,
      snapshotName,
      suiteName,
      testName,
    });

    return await this.visualSnapshotsApi.uploadImageAndCreateSnapshot({
      buildId,
      snapshot: page,
      snapshotName,
      suiteName,
      testName,
      logger,
    });
  }
}
