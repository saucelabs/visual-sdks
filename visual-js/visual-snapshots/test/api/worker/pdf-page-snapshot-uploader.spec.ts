import { DiffingMethod, VisualApi } from "@saucelabs/visual";
import { PdfPageSnapshotUploader } from "../../../src/api/worker/pdf-page-snapshot-uploader.js";
import { LoadedPdfFile } from "../../../src/app/pdf-file.js";
import { PdfFileLoader } from "../../../src/app/pdf-file-loader.js";
import path from "path";

class TestLoadedPdfFile implements LoadedPdfFile {
  constructor(public readonly path: string, public readonly pages: number) {}

  public async getPage(page: number): Promise<Buffer> {
    return Buffer.from(`fake-image-buffer-${page}`);
  }
}

function createUploadId(content: Buffer) {
  return `upload-id:${content.toString("utf-8")}`;
}

describe("PdfPageSnapshotUploader", () => {
  describe("uploadPageSnapshot", () => {
    const consoleInfoSpy = jest
      .spyOn(console, "info")
      .mockImplementation(() => undefined);

    const uploadSnapshotMock = jest.fn<
      ReturnType<VisualApi["uploadSnapshot"]>,
      Parameters<VisualApi["uploadSnapshot"]>
    >();
    const createSnapshotMock = jest.fn<
      ReturnType<VisualApi["createSnapshot"]>,
      Parameters<VisualApi["createSnapshot"]>
    >();

    const visualApiMock = {
      uploadSnapshot: uploadSnapshotMock,
      createSnapshot: createSnapshotMock,
    } as never as VisualApi;

    const files = [
      new TestLoadedPdfFile("file1.pdf", 4),
      new TestLoadedPdfFile("file2.pdf", 5),
    ];

    const loadPdfFileMock = jest
      .fn<
        ReturnType<PdfFileLoader["loadPdfFile"]>,
        Parameters<PdfFileLoader["loadPdfFile"]>
      >()
      .mockImplementation(async (path) => {
        const file = files.find((f) => f.path === path);
        if (!file) {
          throw new Error("file not found");
        }
        return file;
      });

    const pdfLoaderMock: PdfFileLoader = {
      loadPdfFile: loadPdfFileMock,
    };

    const uploader = new PdfPageSnapshotUploader(visualApiMock, pdfLoaderMock);

    beforeEach(() => {
      uploadSnapshotMock.mockReset();
      uploadSnapshotMock.mockImplementation(({ image }) =>
        image && "data" in image
          ? Promise.resolve(createUploadId(image.data))
          : Promise.reject(new Error("image data is missing"))
      );

      createSnapshotMock.mockReset();
      consoleInfoSpy.mockReset();
    });

    it("should call uploadSnapshot visual API", async () => {
      await uploader.uploadPageSnapshot(
        "build-id",
        files[0].path,
        1,
        "suiteName",
        "testName-{filename}",
        "snapshotName-{filename}-{page}"
      );

      expect(uploadSnapshotMock).toHaveBeenCalledWith({
        buildId: "build-id",
        image: { data: await files[0].getPage(1) },
      });
    });

    it("should call createSnapshot visual API", async () => {
      await uploader.uploadPageSnapshot(
        "build-id",
        files[0].path,
        1,
        "suiteName",
        "testName-{filename}",
        "snapshotName-{filename}-{page}"
      );

      expect(createSnapshotMock).toHaveBeenCalledWith({
        buildId: "build-id",
        uploadId: createUploadId(await files[0].getPage(1)),
        name: `snapshotName-${path.basename(files[0].path)}-1`,
        suiteName: "suiteName",
        testName: `testName-${path.basename(files[0].path)}`,
        diffingMethod: DiffingMethod.Balanced,
      });
    });

    it("should print output matching snapshot", async () => {
      await uploader.uploadPageSnapshot(
        "build-id",
        files[0].path,
        1,
        "suiteName",
        "testName-{filename}",
        "snapshotName-{filename}-{page}"
      );

      expect(consoleInfoSpy).toMatchSnapshot();
    });
  });
});
