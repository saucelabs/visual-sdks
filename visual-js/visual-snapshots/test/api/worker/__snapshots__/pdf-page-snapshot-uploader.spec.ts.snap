// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PdfPageSnapshotUploader uploadPageSnapshot should print output matching snapshot 1`] = `
[MockFunction] {
  "calls": [
    [
      "[file1.pdf:1] Uploaded image to build build-id: upload id=upload-id:fake-image-buffer-1.",
    ],
    [
      "[file1.pdf:1] Created a snapshot snapshotName-file1.pdf-1 for build build-id.",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
